File: LeetCode75_Artur/Graphs - DFS/547_Number_of_Provinces.py
"""
https://leetcode.com/problems/number-of-provinces/description/?envType=study-plan-v2&envId=leetcode-75
"""

from typing import List
from collections import defaultdict

class Solution:
    def __init__(self) -> None:
        self.n = 0
        self.visited = []
        self.graph = defaultdict(list)
    
    def dfs(self, root: int):
        self.visited[root] = 1
        
        for neigh in self.graph[root]:
            if not self.visited[neigh]:
                self.dfs(neigh)
    
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        self.n = len(isConnected)
        self.visited = [0 for _ in range(self.n)]
        
        for i in range(self.n):
            for j in range(self.n):
                if isConnected[i][j]:
                    self.graph[i].append(j)
        
        provincesNum = 0
        for i in range(self.n):
            if not self.visited[i]:
                provincesNum += 1
                self.dfs(i)
            
        return provincesNum


if __name__ == "__main__":
    test1 = [[1,1,0],[1,1,0],[0,0,1]]
    test2 = [[1,0,0],[0,1,0],[0,0,1]]
    
    print(Solution().findCircleNum(test1))    
    print(Solution().findCircleNum(test2))    

File: LeetCode75_Artur/Graphs - DFS/547_Number_of_Provinces.cpp
/*
 https://leetcode.com/problems/number-of-provinces/description/?envType=study-plan-v2&envId=leetcode-75
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution
{
private:
    vector<bool> visited;
    int n;
public:
    void dfs(int root, vector<vector<int>>& isConnected)
    {
        for(int i = 0; i < n; ++i)
        {
            if( isConnected[root][i] && !visited[i] )
            {
                visited[i] = true;
                dfs(i, isConnected);
            }
        }
    }

    int findCircleNum(vector<vector<int>>& isConnected)
    {
        n = (int) isConnected.size();
        visited = vector<bool>(n, false);
        int provincesNum = 0;
        for(int i = 0; i < n; ++i)
        {
            if( !visited[i] )
            {
                visited[i] = true;
                dfs(i, isConnected);
                provincesNum++;
            }
        }

        return provincesNum;
    }
};


int main()
{
    Solution solution;

    vector<vector<int>> test1 = {{1,1,0},{1,1,0},{0,0,1}};
    vector<vector<int>> test2 = {{1,0,0},{0,1,0},{0,0,1}};

    cout << solution.findCircleNum(test1) << endl;
    cout << solution.findCircleNum(test2) << endl;
}

/*
 Main idea:
 В загальній функції запускаємо dfs для всіх непройдених
 вершин і рахуємо кількість цих запусків, тоді отримуємо
 загальну кількість провінцій.

 Часова складність: O(n^2)
 Для кожної вершини перевіряється суміжність і відвіданість
 інших вершин.
*/